onnxruntime/python/tools/quantization/calibrate.py

```diff
@@ -226,12 +239,242 @@ def compute_range(self):
+
+    ref: https://github.com//apache/incubator-mxnet/blob/master/python/mxnet/contrib/quantization.py
+    """
+    def __init__(self, num_quantized_bins=128):
+        self.histogram_dict = {}
+        self.num_quantized_bins= num_quantized_bins
+
+    def get_histogram_dict(self):
+        return self.histogram_dict
+
+    def collect(self, name_to_arr):
+        for tensor, data_arr in name_to_arr.items():
+            data_arr = np.asarray(data_arr)
+            data_arr = data_arr.flatten()
+
+            if data_arr.size > 0:
+                min_value = np.min(data_arr)
+                max_value = np.max(data_arr)
+            else:
+                min_value = 0
+                max_value = 0
+
+            threshold = max(abs(min_value), abs(max_value))
+
+            if tensor in self.histogram_dict:
+                old_histogram = self.histogram_dict[tensor]
+                self.histogram_dict[tensor] = self.merge_histogram(old_histogram, data_arr, min_value, max_value, threshold)
+            else:
+                # hist, hist_edges = np.histogram(data_arr, self.num_quantized_bins, range=(min_value, max_value))
+                hist, hist_edges = np.histogram(data_arr, self.num_quantized_bins, range=(-threshold, threshold))
+                self.histogram_dict[tensor] = (hist, hist_edges, min_value, max_value, threshold)
+
+    def merge_histogram(self, old_histogram, data_arr, new_min, new_max, new_threshold):
+
+        (old_hist, old_hist_edges, old_min, old_max, old_threshold) = old_histogram
+
+        if new_threshold <= old_threshold:
+            new_hist, _ = np.histogram(data_arr, len(old_hist), range=(-old_threshold, old_threshold))
+            return (new_hist + old_hist, old_hist_edges, min(old_min, new_min), max(old_max, new_max), old_threshold)
+        else:
+            if old_threshold == 0:
+                hist, hist_edges = np.histogram(data_arr, new_num_bins, range=(-new_threshold, new_threshold))
+                hist[len(hist) // 2] += len(old_hist)
+            else:
+                old_num_bins = len(old_hist)
+                old_stride = 2 * old_threshold / old_num_bins
+                half_increased_bins = int((new_threshold - old_threshold) // old_stride + 1) 
+                new_num_bins = old_num_bins + 2 * half_increased_bins
+                new_threshold = half_increased_bins * old_stride + old_threshold
+                hist, hist_edges = np.histogram(data_arr, new_num_bins, range=(-new_threshold, new_threshold))
+                hist[half_increased_bins:new_num_bins-half_increased_bins] += old_hist
+            return (hist, hist_edges, min(old_min, new_min), max(old_max, new_max), new_threshold)
+
+    def get_optimal_collection_result(self):
+        return self.get_optimal_thresholds(self.histogram_dict, self.num_quantized_bins)
+
+
+    def get_optimal_thresholds(self, histogram_dict, num_quantized_bins=128):
```

Comment:
Is it more simple to merge them together?

Reply:
I've merged the them.